### PR TITLE
Customizable error page buffer size

### PR DESCRIPTION
**Pull request by dkaukov**
Ability to increase buffer size past 8k in the custom `ErrorHandler`

### CODE REVIEW

**Comment by joakime** [comment]
The HttpConfiguration is how you control the output buffer size.

What is the use case for this?

**Comment by dkaukov** [comment]
True, but
```
    bufferSize = Math.min(8192, bufferSize); // TODO ?
```
limiting it to the 8k. We need a way to increase  the limit.

**Comment by joakime** [comment]
> limiting it to the 8k. We need a way to increase the limit.

I understand that you want to increase the buffer size limit, but why?
What's the use case for a larger buffer size? (I ask, as there's probably several different ways to accomplish your use case without increasing this specific buffer size)

**Comment by dkaukov** [comment]
Hi, In the 
```
@Override
protected void writeErrorHtml(Request request, Writer writer, Charset charset, int code, String message, Throwable cause, boolean showStacks) throws IOException {
  ....
}
```
we are generating custom error page using Writer. Something like:
```
  themeLeafTemplate.execute(writer, errorInfo);
```
If error page is bigger than buffer size it is just trimmed.

**Comment by joakime** [comment]
Need to dig into why we are doing this via a static sized buffer.

I would like to see if this need can be accomplished via a Writer created with the Content.Sink features instead. (will try this in a different branch soon)

**Comment by dkaukov** [comment]
Hi @joakime , maybe we can merge this one in a meantime?

**Comment by gregw** [comment]
> Need to dig into why we are doing this via a static sized buffer.

I believe the reasoning is that we  don't want to get into complex responses when generating error pages, as if we are in an error state, we want simplicity.... otherwise we can get errors within errors within errors.....

I'll review this PR for merging now...

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ErrorHandler.java [hunk 1]
```diff
@@ -262,6 +261,13 @@ else if (charsets.contains(StandardCharsets.ISO_8859_1))
         }
     }
 
+    protected int getBufferSize(Request request)
+    {
+        int bufferSize = request.getConnectionMetaData().getHttpConfiguration().getOutputBufferSize();
+        bufferSize = Math.min(8192, bufferSize); // TODO ?
```
Comment (gregw) [inline jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ErrorHandler.java]: Let's remove the TODO from the code

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ErrorHandler.java [hunk 2]
```diff
@@ -262,6 +261,13 @@ else if (charsets.contains(StandardCharsets.ISO_8859_1))
         }
     }
 
+    protected int getBufferSize(Request request)
```
Comment (gregw) [inline jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ErrorHandler.java]: We need a bit of javadoc to say that this is getting the buffer size for the entire error response (any larger responses will be truncated).   Then say that default implementation imposes an 8K limit on the standard buffer size, but that implementations may override the method to provide some other size.

**Review by gregw — CHANGES_REQUESTED** [review]
This is OK.... but I think I'd prefer a simple `setBufferSize` method, with a -1 meaning that a heuristic will be used (e.g. min(8K, httpConfig.bufferSize)).  Default value will be -1 and `public int getBufferSize()` would return -1 (or the set value).  Then have a `protect int computeBufferSize(Request request)` that would check for -1 and do the min(8k, config.bufferSize) thang.

**Comment by gregw** [comment]
> This is OK.... but I think I'd prefer a simple `setBufferSize` method, with a -1 meaning that a heuristic will be used (e.g. min(8K, httpConfig.bufferSize)). Default value will be -1 and `public int getBufferSize()` would return -1 (or the set value). Then have a `protect int computeBufferSize(Request request)` that would check for -1 and do the min(8k, config.bufferSize) thang.

Note the reason I prefer a setter, is that it can be used from XML.

**Comment by dkaukov** [comment]
@gregw Could you please take a  look now.

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ErrorHandler.java [hunk 3]
```diff
@@ -73,6 +73,7 @@ public class ErrorHandler implements Request.Handler
     boolean _showStacks = false;
     boolean _showCauses = false;
     boolean _showMessageInTitle = true;
+    int _bufferSize = -1;
```
Comment (gregw) [inline jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ErrorHandler.java:78]: Hmmm why are these fields not private?   Not a problem for this PR, but strange....

**Review by gregw — DISMISSED** [review]
This looks good to me!

**Comment by gregw** [comment]
CI failure is a flake.  rerunning

**Comment by janbartel** [comment]
Re-running CI test again.

**Review by gregw — APPROVED** [review]

**Comment by gregw** [comment]
Known CI flake

**Comment by gregw** [comment]
@dkaukov thanks

**Pull request merged**